pkg/cache/cache.go

```diff
@@ -0,0 +1,275 @@
+package cache
+
+import (
+	glog "github.com/Sirupsen/logrus"
+	"github.com/kylelemons/godebug/pretty"
+	"github.com/patrickmn/go-cache"
+	calicoClient "github.com/projectcalico/libcalico-go/lib/client"
+	k8sCache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"reflect"
+	"time"
+)
+
+// ResourceCache stores calico representation of kubernetes objects.
+// It tries to be always in sync with kubernetes cache with the
+// help of reconcilor. Adds keys of internal objects to workqueue
+// if any modifications are done to them. Controller will further
+// sync these keys to calico ETCD datastore.
+type ResourceCache interface {
+
+	// Sets the key to the provided value, and generates an update
+	// on the queue the value has changed.
+	Set(key string, value interface{})
+
+	// Gets the value associated with the given key.  Returns nil
+	// if the key is not present.
+	Get(key string) (interface{}, bool)
+
+	// Sets the key to the provided value, but does not generate
+	// and update on the queue ever.
+	Prime(key string, value interface{})
+
+	// Deletes the value identified by the given key from the cache, and
+	// generates an update on the queue if a value was deleted.
+	Delete(key string)
+
+	// Lists the keys currently in the cache.
+	ListKeys() []string
+
+	// Starts the cache.
+	Run(stopChan chan struct{}, reconcilerPeriod string)
+
+	// Get workqueue
+	GetQueue() workqueue.RateLimitingInterface
+}
+
+// ResourceCacheArgs struct passed to constructor of ResourceCache.
+// Groups togather all the arguments to pass in single struct.
+type ResourceCacheArgs struct {
+	// ListFunc returns a list of objects.  Responsible for filtering any
+	// objects which should not be monitored by the cache / controller.
+	ListFunc func() ([]interface{}, error)
+
+	// Takes an object and returns the key string which identifies it in the cache.
+	// e.g. namespace.name
+	KeyFunc func(obj interface{}) string
+
+	// The channel on which modified keys will be sent, if given.
+	OutChan chan string
+
+	// Calico Client
+	Client *calicoClient.Client
+
+	// Type of object cache will hold
+	// Set() API will verfiy the object type before storing it in cache
+	ObjectType reflect.Type
+}
+
+// CalicoCache implements ResourceCache interface
+// Cache only stores pointer to objects instead of actual objects
```

Comment:
Updated Cache implementation to store calico objects rather than points. DeepEqual() is working fine now. Will use DeepEqual() instead of pretty package.